api/v1alpha1/package_webhook.go

```diff
@@ -0,0 +1,206 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/xeipuuv/gojsonschema"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/yaml"
+)
+
+type packageValidator struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+func InitPackageValidator(mgr ctrl.Manager) error {
+	mgr.GetWebhookServer().
+		Register("/validate-packages-eks-amazonaws-com-v1alpha1-package",
+			&webhook.Admission{Handler: &packageValidator{
+				Client: mgr.GetClient(),
+			}})
+	return nil
+}
+
+func (v *packageValidator) Handle(ctx context.Context, request admission.Request) admission.Response {
+	p := &Package{}
+	err := v.decoder.Decode(request, p)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("decoding request: %w", err))
+	}
+
+	bundles, err := v.listBundles(ctx)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("listing package bundles: %w", err))
+	}
+
+	pbc, err := v.getActiveController(ctx)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("getting PackageBundleController: %v", err))
+	}
+
+	activeBundle := v.getActiveBundle(bundles, pbc.Spec.ActiveBundle)
+
+	isConfigValid, err := v.isPackageConfigValid(p, &activeBundle)
+
+	resp := &admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{Allowed: isConfigValid},
+	}
+
+	if !isConfigValid {
+		msg := fmt.Sprintf("package %s failed validation with error: %v", p.Name, err)
+		resp.AdmissionResponse.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Code:    http.StatusBadRequest,
+			Message: msg,
+			Reason:  metav1.StatusReasonBadRequest,
+		}
+	}
+
+	return *resp
+}
+
+func (v *packageValidator) getActiveBundle(bundles *PackageBundleList, b string) PackageBundle {
+	var activeBundle PackageBundle
+	for _, bundle := range bundles.Items {
+		if bundle.Name == b {
+			activeBundle = bundle
+		}
+	}
+	return activeBundle
+}
+
+func (v *packageValidator) getActiveController(ctx context.Context) (PackageBundleController, error) {
```

Comment:
`getActiveController` and `getActiveBundle`/`getBundle` should be part of a more structured API for all our resources. I'd be fine if we consider this out-of-scope for this PR but we need to get started on that as it will help with testing as well.

Reply:
There is actually an api that already exists for it but including it in my code causes an import cycle which can't be solved without a huge revamp of the codebase. We can rethink of this in a more structured manner in a later PR.